Fixed unchecked local variables safercpp errors in UserMediaController.cpp

Phinny01 opened the pull request:
#### 3ba158e16f44332473094ce255d8fed3b90bf494
<pre>
Fixed unchecked local variables safercpp errors in UserMediaController.cpp
<a href="https://rdar.apple.com/155265597">rdar://155265597</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295552">https://bugs.webkit.org/show_bug.cgi?id=295552</a>

Reviewed by Chris Dumez and Jean-Yves Avenard.

Deleted the destructors in UserMediaController and WebUserMediaClient as they were unnecessary.
Moved the create method implementation into the body of the WebUserMediaClient class. Marked
WebUserMediaClient as final to prevent further subclassing.

* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::~UserMediaController): Deleted.
* Source/WebCore/Modules/mediastream/UserMediaController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
(WebKit::WebUserMediaClient::create): Deleted.

Canonical link: <a href="https://commits.webkit.org/297277@main">https://commits.webkit.org/297277@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ccb35603f2bf828980da974444498a07d010315

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30704 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21135 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117068 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61305 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113000 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84413 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64859 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24416 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18111 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60884 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18176 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119953 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28302 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93375 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93199 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23763 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38257 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15994 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34068 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37976 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43452 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37640 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40974 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -45,14 +45,15 @@ ASCIILiteral UserMediaController::supplementName()
 UserMediaController::~UserMediaController()
 {
-    m_client->pageDestroyed();
+    if (RefPtr<UserMediaClient> mediaClient = m_client.get())
```
```suggestion
    if (RefPtr mediaClient = m_client.get())
```
let template deduction work.

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -45,14 +45,15 @@ ASCIILiteral UserMediaController::supplementName()
 }
 
-UserMediaController::UserMediaController(UserMediaClient* client)
+UserMediaController::UserMediaController(const RefPtr<UserMediaClient>& client)
```
Can it ever be called with `nullptr`? could we make this a reference instead?

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -107,7 +108,8 @@ void UserMediaController::checkDocumentForVoiceActivity(const Document* document
 
     m_shouldListenToVoiceActivity = shouldListenToVoiceActivity;
-    m_client->setShouldListenToVoiceActivity(m_shouldListenToVoiceActivity);
+    if (RefPtr<UserMediaClient> mediaClient = m_client.get())
```
```suggestion
    if (RefPtr mediaClient = m_client.get())
```

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -68,41 +68,48 @@ class UserMediaController : public Supplement<Page> {
 inline void UserMediaController::requestUserMediaAccess(UserMediaRequest& request)
 {
-    m_client->requestUserMediaAccess(request);
+    if (RefPtr<UserMediaClient> mediaClient = m_client.get())
```
ditto

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaClient.h:
```diff
@@ -47,7 +48,7 @@ class UserMediaRequest;
 struct MediaDeviceHashSalts;
 
-class UserMediaClient {
+class UserMediaClient : public CanMakeWeakPtr<UserMediaClient>, public RefCounted<UserMediaClient> {
```
```suggested
class UserMediaClient : public RefCountedAndCanMakeWeakPtr<UserMediaClient> {
```

jyavenard reviewed (changes requested):
If you make your object RefCounted, you need to change how your object is allocated using a create method, otherwise you will get a leak

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaClient.h:
```diff
@@ -65,11 +66,15 @@ class UserMediaClient {
     virtual void updateCaptureState(const Document&, bool isActive, MediaProducerMediaCaptureKind, CompletionHandler<void(std::optional<Exception>&&)>&&) = 0;
     virtual void setShouldListenToVoiceActivity(bool) = 0;
 
+
```
don't need two blank lines.

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -107,14 +108,17 @@ void UserMediaController::checkDocumentForVoiceActivity(const Document* document
 #if ENABLE(MEDIA_SESSION)
-    for (auto& document : m_voiceActivityDocuments)
-        Ref(document)->voiceActivityDetected();
+    for (auto& document : m_voiceActivityDocuments) {
```
```suggested
    for (RefPtr document : m_voiceActivityDocuments)
```

and then you can remove the `if`

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -107,14 +108,17 @@ void UserMediaController::checkDocumentForVoiceActivity(const Document* document
-        Ref(document)->voiceActivityDetected();
+    for (auto& document : m_voiceActivityDocuments) {
+        if (RefPtr protectedDocument = &document)
+            protectedDocument->voiceActivityDetected();
```
this will always be true. there's no point testing for it.

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -41,10 +41,10 @@ class UserMediaRequest;
     ~UserMediaController();
 
-    UserMediaClient* client() const { return m_client; }
+    RefPtr<UserMediaClient> client() const { return m_client.get(); }
```
```suggested
    RefPtr<UserMediaClient> protectedClient() const { return m_client.get(); }
```

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -68,41 +68,48 @@ class UserMediaController : public Supplement<Page> {
 inline void UserMediaController::requestUserMediaAccess(UserMediaRequest& request)
 {
-    m_client->requestUserMediaAccess(request);
+    if (RefPtr mediaClient = m_client.get())
```
```suggested
    if (RefPtr client = protectedClient())
```

cdumez replied:
Actually, our style is:
```
if (RefPtr client = m_client.get())
```
We never use `protectedFoo()` functions to initialize local variables since the name is longer and there is no perf /safety benefit to be. Please do not apply this review feedback.

jyavenard commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
```diff
@@ -33,9 +33,12 @@ class WebPage;
     WTF_MAKE_TZONE_ALLOCATED(WebUserMediaClient);
 public:
-    WebUserMediaClient(WebPage&);
+    static Ref<WebUserMediaClient> create(WebPage&);
```
can make this method inline maybe.

jyavenard reviewed (approved):
LGTM, some nits 

Commit log needs updating

cdumez commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -45,19 +45,20 @@ ASCIILiteral UserMediaController::supplementName()
+void provideUserMediaTo(Page* page, Ref<UserMediaClient>&& client)
 {
-    UserMediaController::provideTo(page, UserMediaController::supplementName(), makeUnique<UserMediaController>(client));
+    UserMediaController::provideTo(page, UserMediaController::supplementName(), makeUnique<UserMediaController>(WTFMove(client)));
```
`WTFMove()` is unnecessary here.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.cpp:
```diff
@@ -40,14 +40,19 @@ WebUserMediaClient::WebUserMediaClient(WebPage& page)
 void WebUserMediaClient::pageDestroyed()
 {
-    delete this;
+    // RefCounted object will handle deallocation automatically
```
That's only true if UserMediaController clears its m_client pointer after calling `pageDestroyed()`. I do not think you currently do this in this PR so `WebUserMediaClient` definitely no longer gets deallocated at the same time as before at least, which could be a problem.

Also, I would omit this comment.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
```diff
@@ -33,9 +33,12 @@ class WebPage;
+    static Ref<WebUserMediaClient> create(WebPage&);
     ~WebUserMediaClient() { }
 
+private:
```
There is already a private section in this class, please use the existing one.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/329079039ba586a557950887636aa210b6406861. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47692

cdumez commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -45,19 +45,20 @@ ASCIILiteral UserMediaController::supplementName()
 
-UserMediaController::UserMediaController(UserMediaClient* client)
-    : m_client(client)
+UserMediaController::UserMediaController(Ref<UserMediaClient> client)
```
`Ref<UserMediaClient>&&`

cdumez commented on Source/WebCore/Modules/mediastream/UserMediaController.cpp:
```diff
@@ -45,19 +45,19 @@ ASCIILiteral UserMediaController::supplementName()
 UserMediaController::~UserMediaController()
 {
-    m_client->pageDestroyed();
+    m_client = nullptr;
```
It is unnecessary to null out data members in destructors.

Phinny01 replied:
I removed the destructor now

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
```diff
@@ -30,16 +30,17 @@ namespace WebKit {
 class WebPage;
 
-class WebUserMediaClient : public WebCore::UserMediaClient {
+class WebUserMediaClient : public WebCore::UserMediaClient, public RefCounted<WebUserMediaClient> {
```
nit: class can be marked as `final` I think.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
```diff
@@ -30,16 +30,17 @@ namespace WebKit {
-    WebUserMediaClient(WebPage&);
-    ~WebUserMediaClient() { }
+    static Ref<WebUserMediaClient> create(WebPage&);
+    ~WebUserMediaClient() override { }
```
we don't usually mark destructors as override. Since this destructor does nothing, I would just omit it.

Phinny01 replied:
I have omitted this

cdumez reviewed (approved):
LGTM with nits.

jyavenard commented on Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
```diff
@@ -53,10 +54,15 @@ class WebUserMediaClient : public WebCore::UserMediaClient {
 };
 
-} // namespace WebCore
+inline Ref<WebUserMediaClient> WebUserMediaClient::create(WebPage& page)
```
why not make this directly in the class definition ?

Phinny01 replied:
I have addressed this

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1ccb35603f2bf828980da974444498a07d010315. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47692

Phinny01 committed:
Fixed unchecked local variables safercpp errors in UserMediaController.cpp
rdar://155265597
https://bugs.webkit.org/show_bug.cgi?id=295552

Reviewed by Chris Dumez and Jean-Yves Avenard.

Deleted the destructors in UserMediaController and WebUserMediaClient as they were unnecessary.
Moved the create method implementation into the body of the WebUserMediaClient class. Marked
WebUserMediaClient as final to prevent further subclassing.

* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
(WebCore::UserMediaController::~UserMediaController): Deleted.
* Source/WebCore/Modules/mediastream/UserMediaController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:
(WebKit::WebUserMediaClient::create): Deleted.

Canonical link: https://commits.webkit.org/297277@main
